src/hardware/pcspeaker_impulse.cpp

```diff
@@ -0,0 +1,543 @@
+	for (auto k = 1; k < SINC_ACCURACY; ++k) {
+		result *= cos(t / pow(2.0, k));
+	}
+	return result;
+}
+
+float PcSpeakerImpulse::CalcImpulse(const double t)
+{
+	// raised-cosine-windowed sinc function
+	const double fs = sample_rate;
+	const auto fc   = fs / (2 + static_cast<double>(cutoff_margin));
+	const auto q    = static_cast<double>(filter_quality);
+	if ((0 < t) && (t * fs < q)) {
+		const auto window    = 1.0 + cos(2 * fs * M_PI * (q / (2 * fs) - t) / q);
+		const auto amplitude = window * (sinc(2 * fc * M_PI * (t - q / (2 * fs)))) / 2.0;
+		return static_cast<float>(amplitude);
+	} else
+		return 0.0f;
+}
+
+void PcSpeakerImpulse::AddImpulse(float index, int16_t amplitude)
+{
+	// Did the amplitude change?
+	if (amplitude == pit.prev_amplitude)
+		return;
+
+	pit.prev_amplitude = amplitude;
+
+	// Wake the channel if it was sleeping
+	if (!channel->is_enabled)
+		channel->Enable(true);
+
+	// Make sure the time index is valid
+	index = clamp(index, 0.0f, 1.0f);
+#ifndef REFERENCE
+	const auto samples_in_impulse = index * sample_rate_per_ms;
+	auto phase = static_cast<int>(samples_in_impulse * oversampling_factor) % oversampling_factor;
+	auto offset = static_cast<int>(samples_in_impulse);
+	if (phase != 0) {
+		offset++;
+		phase = oversampling_factor - phase;
+	}
+
+	for (uint16_t i = 0; i < filter_quality; ++i) {
+		const auto wave_i    = check_cast<uint16_t>(offset + i);
+		const auto impulse_i = check_cast<uint16_t>(phase + i * oversampling_factor);
+		waveform_deque.at(wave_i) += amplitude * impulse_lut.at(impulse_i);
+	}
+}
+#else
+	const auto portion_of_ms = static_cast < double(index) / 1000.0;
+	for (size_t i = 0; i < waveform_deque.size(); ++i) {
+		const auto impulse_time = static_cast<double>(i) / sample_rate - portion_of_ms;
+
+		waveform_deque[i] += amplitude * CalcImpulse(impulse_time);
+	}
+}
+#endif
+
+void PcSpeakerImpulse::ChannelCallback(uint16_t requested_frames)
+{
+	ForwardPIT(1.0f);
+	pit.last_index = 0;
+
+	constexpr auto num_channels = 2;
+	// Static vectors and samples re-used for all callbacks
+	static std::vector<float> in_buffer(num_channels);
+	static std::vector<int16_t> out_buffer(num_channels);
+	static auto &out_sample = out_buffer.front();
+	static auto &in_sample  = in_buffer.front();
+
+	while (requested_frames > 0 && waveform_deque.size()) {
+		// Pop the first sample off the waveform
+		in_sample += waveform_deque.front();
+		waveform_deque.pop_front();
+		waveform_deque.push_back(0.0f);
+
+		// Soft-limit the sample, if needed
+		soft_limiter->Process(in_buffer, 1, out_buffer);
+
+		// Pass the sample to the mixer and decrement the requested frames
+		channel->AddSamples_m16(1, &out_sample);
+		--requested_frames;
+
+		// Keep a tally of sequential silence so we can sleep the channel
+		tally_of_silence = out_sample ? 0 : tally_of_silence + 1;
+
+		// Scale down the running volume amplitude. Eventually it will
+		// hit 0 if no other waveforms are generated.
+		in_sample *= high_pass_amount;
```

Comment:
High pass? This is just doing volume scaling, no? I think it should be named more appropriately.